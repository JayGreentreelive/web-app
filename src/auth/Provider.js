import React, {
    useReducer,
    useEffect,
    useState,
    useContext,
    createContext
} from 'react'

export const AUTH_TOKEN_KEY = 'auth-token'
const initialState = ''
const localState = localStorage.getItem(AUTH_TOKEN_KEY)
const AuthContext = createContext()
const useAuth = () => useContext(AuthContext)

let reducer = (token, newToken) => {
    if (newToken === null) {
        localStorage.removeItem(AUTH_TOKEN_KEY)
        return initialState
    }
    return newToken || token
}

const AuthProvider = (props) => {
    const [token, setToken] = useReducer(reducer, localState || initialState)
    const [isLoggedIn, setIsLoggedIn] = useState(!!token)
    const logout = () => {
        setToken(null)
    }

    useEffect(() => {
        localStorage.setItem(AUTH_TOKEN_KEY, token)
        setIsLoggedIn(!!token)
    }, [token])

    return (
        <AuthContext.Provider
            value={{
                token,
                setToken,
                isLoggedIn,
                logout
            }}
        >
            {props.children}
        </AuthContext.Provider>
    )
}

export {
    AuthContext,
    AuthProvider,
    useAuth
}
