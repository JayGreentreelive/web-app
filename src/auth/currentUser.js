import { useState, useEffect } from 'react'
import { useLazyQuery } from 'react-apollo'
import { get } from 'lodash'
import { useAuth } from './'
import gql from 'graphql-tag'
import moment from 'moment'

const GET_CURRENT_PERSON = gql`
    query {
        currentUser {
            profile {
                firstName
                lastName 
            }
        }
    }
`

const useCurrentUser = (fields) => {
    const { token, isLoggedIn, logout } = useAuth()
    const [doRefetch, setRefetch] = useState(false)
    const [getCurrentUser, {
        data,
        loading,
        error,
        called,
        refetch
    }] = useLazyQuery(GET_CURRENT_PERSON)
    let currentUser = get(data, 'currentUser.profile', null)

    useEffect(() => {
        if (isLoggedIn) setRefetch(true)
    }, [token])

    try {
        if (doRefetch) {
            // If the state of doRefetch is true,
            //  attempt to refetch the data, then set the
            //  state to false to avoid an infinite loop
            refetch().catch(e => console.log({ e }))
            setRefetch(false)
            currentUser = null // data does not automatically update, so this needs to be set manually
        } else if (error && !doRefetch && !loading) {
            // Log error and log out if there is an error found
            //  Ignore the error if doRefetch is set to true because
            //  the error is left over from a previous request and
            //  ignore errors if the request is still loading
            console.error("Authentication error: logging out")
            logout()
            currentUser = null // data does not automatically update, so this needs to be set manually
        }

        // Request data on initial load:
        //  if the request has not already been called,
        //  check to see if the use is logged in before calling
        if (!called && isLoggedIn) {
            getCurrentUser()
        }
    } catch (e) {
        console.error('GQL error:', { e })
        currentUser = null // data does not automatically update, so this needs to be set manually
    }

    return {
        currentUser,
        loading,
        error
    }
}

export {
    useCurrentUser
}