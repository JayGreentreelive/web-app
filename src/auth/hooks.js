import React, { useState, useEffect } from 'react'

const LOGIN_KEY = 'login-status'
const AUTH_TOKEN_KEY = 'auth-token'

export const useLogin = () => {
    const [isLoggedIn, setIsLoggedIn] = useState(localStorage.getItem(LOGIN_KEY) || false)

    useEffect(() => {
        localStorage.setItem(LOGIN_KEY, isLoggedIn)
    }, [isLoggedIn])

    const setLoginStatus = (status) => setIsLoggedIn(status)

    return { isLoggedIn, setLoginStatus }
}

export const useAuth = () => {
    const [authToken, setAuthToken] = useState(localStorage.getItem(AUTH_TOKEN_KEY))
    const [isLoggedIn, setIsLoggedIn] = useState(authToken !== '')

    useEffect(() => {
        localStorage.setItem(AUTH_TOKEN_KEY, authToken)
        setIsLoggedIn(authToken !== '')
    }, [authToken, setIsLoggedIn])

    return {
        isLoggedIn,
        setAuthToken,
        authToken
    }
}