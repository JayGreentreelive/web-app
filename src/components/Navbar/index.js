export { default } from './Navbar'
export { default as NavbarWithOpacity } from './NavbarWIthOpacity'
