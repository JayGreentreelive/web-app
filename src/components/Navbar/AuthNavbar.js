import React, { useState } from 'react'
import { Query, useQuery } from 'react-apollo'
import PropTypes from 'prop-types'
import gql from 'graphql-tag'
import classnames from 'classnames'
import { toLower, get, has, find, camelCase } from 'lodash'

import { Navbar, Nav, Dropdown } from 'react-bootstrap'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBars } from '@fortawesome/fontawesome-pro-light'
import { Media } from '@christfellowshipchurch/web-ui-kit'

import { useAuth, useAuthQuery } from '../../auth'

const GET_WEBSITE_HEADER = gql`
  query websiteNavigation($website:String!) {
    getWebsiteNavigation(website:$website) {
      images {
        sources {
          uri
        }
        name
      }
    }
  }
`

const GET_PROFILE_IMAGE = gql`
  query {
    currentUser {
      profile {
        firstName
        photo {
          uri
        }
      }
    }
  }
`

// Takes a collection of images from the API's return data and formats
//  it to be an array of the following object structure: { imageKey: { uri, alt } }
const imageArrayToObject = (images) => {
  let imagesObj = {}

  images.forEach((n, i) => {
    const key = camelCase(get(n, 'name', i))
    const uri = get(n, 'sources[0].uri', '')
    const alt = get(n, 'name', 'Christ Fellowship Church')

    imagesObj[key] = { uri, alt }
  })

  return imagesObj
}

const BrandImg = ({
  className,
  uri,
  alt
}) =>
  <Navbar.Brand
    href="/"
    className={classnames(
      "align-self-start",
      className
    )}
  >
    <img
      src={uri}
      style={{ height: '40px', width: 'auto' }}
      alt={alt}
    />
  </Navbar.Brand>

const ProfileConnected = () => {
  const { logout } = useAuth()
  const { loading, error, data } = useAuthQuery(GET_PROFILE_IMAGE)

  if (has(data, 'currentUser.profile')) return (
    <div
      className={classnames(
        'd-flex',
        'align-items-center'
      )}
    >
      <div className="text-right">
        <Dropdown>
          <Dropdown.Toggle
            variant="link"
          >
            {get(data, 'currentUser.profile.firstName')}
          </Dropdown.Toggle>

          <Dropdown.Menu
            className="text-right"
          >
            <Dropdown.Item
              href="#"
              onClick={() => logout()}
            >
              Log Out
            </Dropdown.Item>
          </Dropdown.Menu>
        </Dropdown>
      </div>

      {get(data, 'currentUser.profile.photo.uri', '') !== ''
        ? (
          <div style={{ width: 48, height: 48 }}>
            <Media
              imageUrl={get(data, 'currentUser.profile.photo.uri', '')}
              imageAlt={`Christ Fellowship Church - ${get(data, 'currentUser.profile.firstName')}`}
              ratio="1by1"
              circle
            />
          </div>
        )
        : <i className="fal fa-user-circle fa-2x"></i>
      }
    </div>
  )

  // log out on error

  if (loading) return <i className="fal fa-user-circle fa-2x"></i>
}

const NavbarConnected = ({
  bg,
  variant,
  brandImageKey,
  onToggle,
  fixed,
  links,
}) => {

  const website = process.env.REACT_APP_WEBSITE_KEY
  const { data } = useQuery(GET_WEBSITE_HEADER, {
    variables: { website },
    fetchPolicy: "cache-and-network"
  })

  const images = imageArrayToObject(get(data, 'getWebsiteNavigation.images', []))
  const brandImage = get(images, brandImageKey, null)
  const navbarProps = {
    bg,
    variant,
    expand: 'lg'
  }

  // We use sticky styling as the default so that padding is respected
  //    with the option to override it to use a fixed styling if preferred
  if (fixed) navbarProps.fixed = "top"
  else navbarProps.sticky = "top"

  return (
    <Navbar
      {...navbarProps}
    >
      {brandImage &&
        <BrandImg
          uri={brandImage.uri}
          alt={brandImage.alt}
          className="d-lg-none"
        />
      }

      <Navbar.Toggle
        aria-controls="basic-navbar-nav"
        onClick={onToggle}
        className="border-0">
        <FontAwesomeIcon icon={faBars} size="1x" />
      </Navbar.Toggle>

      <Navbar.Collapse>
        <div
          className={classnames(
            'd-flex',
            'flex-row',
            'justify-content-between',
            "align-items-center",
            'w-100',
            "px-lg-3"
          )}
        >
          <div style={{ flex: 1 }}>
            {brandImage &&
              <BrandImg
                uri={brandImage.uri}
                alt={brandImage.alt}
                className={"d-none d-lg-block"}
              />
            }
          </div>

          <Nav
            style={{ flex: 1 }}
            className={classnames(
              "align-items-center",
              'justify-content-center'
            )}
          >
            {links.map((link, i) => (
              <Nav.Link
                key={i}
                href={link.action}
                className={classnames(
                  'mx-3',
                  'font-weight-normal'
                )}
              >
                {link.call}
              </Nav.Link>
            ))}
          </Nav>

          <div
            style={{ flex: 1 }}
            className={classnames(
              "d-flex",
              "justify-content-end"
            )}
          >
            <ProfileConnected />
          </div>
        </div>
      </Navbar.Collapse>
    </Navbar>
  )
}

NavbarConnected.propTypes = {
  bg: PropTypes.string,
  variant: PropTypes.string,
  brandImageKey: PropTypes.string,
  fixed: PropTypes.bool,
  links: PropTypes.arrayOf(
    PropTypes.shape({
      call: PropTypes.string,
      action: PropTypes.string,
    }))
}

NavbarConnected.defaultProps = {
  bg: 'white',
  variant: 'light',
  brandImageKey: 'brandImage',
  fixed: false,
  links: [
    { call: 'Groups', action: '/groups' },
    { call: 'Serve', action: '/serve' },
    { call: 'Events', action: '/events' },
    { call: 'Give', action: '/give' },
    { call: 'Browse', action: '/articles' },
  ]
}

export default NavbarConnected
