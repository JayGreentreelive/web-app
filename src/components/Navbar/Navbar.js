import React, { useState } from 'react'
import { Query, useQuery } from 'react-apollo'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import { toLower, get, has, find, camelCase } from 'lodash'
import { useScrollPosition } from '../../hooks'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBars } from '@fortawesome/fontawesome-pro-light'

import { Navbar, Nav } from 'react-bootstrap'

import GET_WEBSITE_HEADER from './getWebsiteHeader'
import DefaultIcon from '../../images/default_icon.png'

import { Button } from '../../ui'
import { redirectTo } from '../../utils'
import { useAuth } from '../../auth'
import { AuthNavbar } from './'

// Takes a collection of images from
//  the API's return data and formats
//  it to be an array of the following
//  object structure: { imageKey: { uri, alt } }
const imageArrayToObject = (images) => {
  let imagesObj = {}

  images.forEach((n, i) => {
    const key = camelCase(get(n, 'name', i))
    const uri = get(n, 'sources[0].uri', '')
    const alt = get(n, 'name', 'Christ Fellowship Church')

    imagesObj[key] = { uri, alt }
  })

  return imagesObj
}

const DefaultNavbar = () =>
  <nav className='navbar navbar-default'>
    <a href="/">
      <img
        src={DefaultIcon}
        alt='Christ Fellowship Church Icon'
      />
    </a>
  </nav>

const NavbarConnected = ({
  bg,
  variant,
  brandImageKey,
  onToggle,
  onSelect,
  fixed
}) => {
  const { isLoggedIn, logIn } = useAuth()
  const website = process.env.REACT_APP_WEBSITE_KEY
  const { loading, error, data } = useQuery(GET_WEBSITE_HEADER, {
    variables: { website },
    fetchPolicy: "cache-and-network"
  })

  if (isLoggedIn) return <AuthNavbar />

  // If Query state is loading or there's an error,
  //  return a defaulted header with the CF Icon centered
  if (loading) return <DefaultNavbar />
  if (error) {
    console.error({ error })
    return <DefaultNavbar />
  }

  // Get the data object from the return data or default to null
  const navigationData = get(data, 'getWebsiteNavigation', null)

  if (navigationData) {
    const images = imageArrayToObject(get(navigationData, 'images', []))
    const brandImage = get(images, brandImageKey, null)
    const quickAction = {
      display: has(navigationData, 'quickAction.call') && has(navigationData, 'quickAction.action'),
      call: get(navigationData, 'quickAction.call', ''),
      action: get(navigationData, 'quickAction.action', '')
    }
    const navbarProps = {
      bg,
      variant,
      expand: 'lg',
      collapseOnSelect: true
    }

    // We use sticky styling as the default so that padding is respected
    //    with the option to override it to use a fixed styling if preferred
    if (fixed) navbarProps.fixed = "top"
    else navbarProps.sticky = "top"

    return (
      <Navbar {...navbarProps}>
        {brandImage &&
          <Navbar.Brand href="/">
            <img
              src={brandImage.uri}
              style={{ height: '80px', width: 'auto' }}
              alt={brandImage.alt}
            />
          </Navbar.Brand>
        }
        <Navbar.Toggle
          aria-controls="basic-navbar-nav"
          onClick={onToggle}
          className="border-0">
          <FontAwesomeIcon icon={faBars} size="1x" />
        </Navbar.Toggle>

        <Navbar.Collapse>
          <Nav
            className={classnames(
              "ml-auto",
              "align-items-start",
              "align-items-lg-center",
            )}
          >
            {navigationData.navigationLinks.map((link, i) => (
              <Nav.Link
                key={i}
                href={link.action}
                className='mx-3 my-2'
                onSelect={onSelect}
              >
                {link.call}
              </Nav.Link>
            ))}

            <Nav.Link
              href='#'
              className='mx-3 my-2'
              onSelect={() => {
                onSelect()
                logIn()
              }}
            >
              Log In
            </Nav.Link>

            {quickAction.display &&
              <div className="mx-3 my-2">
                <Button
                  title={quickAction.call}
                  onClick={() => redirectTo(quickAction.action)}
                />
              </div>
            }
          </Nav>
        </Navbar.Collapse>
      </Navbar>
    )
  }

  // If the expected data is not returned from the API,
  //    return the default navbar
  return <DefaultNavbar />
}

NavbarConnected.propTypes = {
  bg: PropTypes.string,
  variant: PropTypes.string,
  brandImageKey: PropTypes.string,
  fixed: PropTypes.bool,
  onToggle: PropTypes.func,
  onSelect: PropTypes.func,
}

NavbarConnected.defaultProps = {
  bg: 'white',
  variant: 'light',
  brandImageKey: 'brandImage',
  fixed: false,
  onToggle: () => true,
  onSelect: () => true,
}

export default NavbarConnected
