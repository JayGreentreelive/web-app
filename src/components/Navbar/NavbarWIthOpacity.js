import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { useScrollPosition } from '../../hooks'
import Navbar from './index'

const NavbarWithOpacity = ({ offset }) => {
  const [bgColor, setBgColor] = useState('transparent')
  const [variant, setVariant] = useState('dark')
  const [imageKey, setImageKey] = useState('brandImageAlt')

  useScrollPosition(({ currPos }) => {
    const opaque = currPos.y > (-1 * offset)

    setBgColor(opaque ? 'transparent' : 'white')
    setVariant(opaque ? 'dark' : 'light')
    setImageKey(opaque ? 'brandImageAlt' : 'brandImage')
  })

  return <Navbar bg={bgColor} variant={variant} brandImageKey={imageKey} />

}

NavbarWithOpacity.propTypes = {
  offset: PropTypes.number
}

NavbarWithOpacity.defaultProps = {
  offset: 50
}

export default NavbarWithOpacity
