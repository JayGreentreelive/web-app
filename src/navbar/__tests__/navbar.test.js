import React from 'react'
import Navbar, { NavbarWithOpacity } from '../index'
import ShallowRenderer from 'react-test-renderer/shallow'
import { AuthProvider } from '../../auth'


describe('Navbar', () => {
    it('renders without crashing', () => {
        const renderer = new ShallowRenderer()
        renderer.render(
            <AuthProvider>
                <Navbar />
            </AuthProvider>
        )
        const tree = renderer.getRenderOutput()

        expect(tree).toMatchSnapshot();
    })
})

describe('NavbarWithOpacity', () => {
    it('renders without crashing', () => {
        const renderer = new ShallowRenderer()
        renderer.render(
            <AuthProvider>
                <NavbarWithOpacity />
            </AuthProvider>
        )
        const tree = renderer.getRenderOutput()

        expect(tree).toMatchSnapshot();
    })
})