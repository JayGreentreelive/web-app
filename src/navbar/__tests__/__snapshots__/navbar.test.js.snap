// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavbarWithOpacity renders without crashing 1`] = `
<NavbarConnected
  bg="transparent"
  brandImageKey="brandImageAlt"
  variant="dark"
/>
`;
